### DISCUSSION TITLE
Inconsistent key inputs between terminal apps on Mac (Terminal and iTerm2)

### DISCUSSION THREAD

**Discussion by blackmann** [charmbracelet/bubbletea#291] [Q&A] [answered]
I'm confident they may be having separate underlying implementations on how key inputs are handled but I don't know how to [think or] approach it as a TUI app.

For my particular case, I have set up a keybind like this:

```go
func NewViewport() Model {
	tabs := []string{"Headers (q)", "Request Body (w)", "Response (e)"}
	keybinds := keymap{
		nextTab:     key.NewBinding(key.WithKeys("alt+f"), key.WithHelp("⌥→", "Next tab")),
		previousTab: key.NewBinding(key.WithKeys("alt+b"), key.WithHelp("⌥←", "Prev tab")),
	}

	return Model{
		tabs:     tabs,
		keybinds: keybinds,
	}
}
```

So `alt+f` for example is used for switching to the next tab but if I enter `option+forward` (which I assume is alt+f in bubbletea land) in iTerm and log the key, I get 

> Key is  alt+

"Key is" is my logging prefix. This effectively allows the switch case to not match my binding.  However on iTerm, I get `alt+f`.

What's the general approach for such cases apart from considering alternate keybinds?

As a secondary request, is there a list/file of all the keys in their string representation for reference for bubbletea?

**Comment by meowgorithm** [comment, marked as answer]
Hi! So terminals don't have a notion of the <kbd>option</kbd> key. On macOS Terminal and iTerm2 you need to tick a box in settings to have option send `alt` (which is often called `meta`). What it sounds like is happening is that one of your terminals is sending `alt`/`meta` from the  option key and the other is not.

On a personal note, when I'm on a Mac I setup <a href="https://sw.kovidgoyal.net/kitty/">Kitty</a> so that I my left <kbd>option</kbd> key is bound to alt/meta while the right <kbd>option</kbd> remains bound to option so I still have access to the system-wide, option-based keybindings—particularly those that let me enter special characters.

* * *

Separately, per your question about string representations for keys, there's no official reference for string representations of keys at the moment, but you can get an idea what they look like below. Not listed are basic latin characters like `"a"`, `"b"`, and so on as well as `" "` which is the literal space key. Also keep in mind that many of these can receive an `"alt+"` modifier, but not all because of limitations of terminal emulators for a variety of reasons.

https://github.com/charmbracelet/bubbletea/blob/515ef02733e61f45428ab16582901e7823f81b1f/key.go#L225-L293

**Comment by blackmann** [comment]
Thanks @meowgorithm. This answer is helpful.
